File: chapter_4_trees_graphs/python/4.4-check_balanced.py
TreeNode = __import__('tree').TreeNode


def check_balanced(root):
    """
    Determines if a binary tree is balanced
    """
    return check_height(root) != float("-inf")


def check_height(root):
    """
    Calculates differences between heights of left and right nodes.
    If difference is greater than 1, return INT_MIN */
    """
    if root == None:
        return -1
    left_height = check_height(root.left)
    if left_height == float("-inf"):
        return float("-inf")
    right_height = check_height(root.right)
    if right_height == float("-inf"):
        return float("-inf")
    difference = left_height - right_height
    if abs(difference) > 1:
        return float("-inf")
    return max(left_height, right_height) + 1


n1 = TreeNode(1)
n2 = TreeNode(2)
n1.left = n2
n3 = TreeNode(3)
n1.right = n3
n4 = TreeNode(4)
n2.left = n4
n5 = TreeNode(5)
n2.right = n5
n6 = TreeNode(6)
n7 = TreeNode(7)
n3.left = n6
# n3.right = n7 # will make it balanced
n8 = TreeNode(8)
n9 = TreeNode(9)
n4.left = n8
n4.right = n9
n10 = TreeNode(10)
n11 = TreeNode(11)
n5.left = n10
n5.right = n11
n12 = TreeNode(12)
n13 = TreeNode(13)
n6.left = n12
n6.right = n13
n15 = TreeNode(15)

print(check_balanced(n1))  # false

File: chapter_4_trees_graphs/cpp/4.4-check_balanced.cpp
#include "tree.h"
#include <climits>

/* Calculates differences between heights of left and right nodes.
 * If difference is greater than 1, return INT_MIN */
int check_height(TreeNode *root)
{
    int left_height, right_height, difference;

    if (!root)
        return -1;
    left_height = check_height(root->left);
    if (left_height == INT_MIN)
        return INT_MIN;
    right_height = check_height(root->right);
    if (right_height == INT_MIN)
        return INT_MIN;
    difference = left_height - right_height;
    if (abs(difference) > 1)
        return INT_MIN;
    return max(left_height, right_height) + 1;
}

/* Determines if a binary tree is balanced */
bool check_balanced(TreeNode *root)
{
    return check_height(root) != INT_MIN;
}

int main()
{
    TreeNode *t1 = new TreeNode(1);
    TreeNode *t2 = new TreeNode(2);
    TreeNode *t3 = new TreeNode(3);
    t1->left = t2;
    t1->right = t3;
    TreeNode *t4 = new TreeNode(4);
    TreeNode *t5 = new TreeNode(5);
    t2->left = t4;
    t2->right = t5;
    TreeNode *t6 = new TreeNode(6);
    TreeNode *t7 = new TreeNode(7);
    t3->left = t6;
    t3->right = t7;

    cout << check_balanced(t1) << endl; // 1

    TreeNode *t8 = new TreeNode(8);
    TreeNode *t9 = new TreeNode(9);
    t6->right = t8;
    t8->right = t9;

    cout << check_balanced(t1) << endl; // 0

    return 0;
}
